### PR TITLE
Vue3: Support multiple setup functions

### PR DESCRIPTION
**Pull request by chakAs3**
Closes #

vue 3 single use of setup() which is causing issue for add-on authors NickMcBurney/storybook-vue3-router#38

## What I did

change the implementation of setup function mostly called in preview.js or plugins , now you can use as many as you want and it won't be overridden by any plugin use

this PR should be merged after merging #21956

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Review by chakAs3 — COMMENTED** [review]
i don't like !, or ts-ignore but sometime i can't find solution so that it will be my last solution to fix  type check

File: docs/snippets/common/storybook-upgrade-prerelease.pnpm.js.mdx
Comment (JReinhold) [inline docs/snippets/common/storybook-upgrade-prerelease.pnpm.js.mdx:2]: `pnpx` has been deprecated in favor of `pnpm dlx`. You can't really see it in the docs anymore, because `pnpx` hasn't been mentioned since v6, whose docs are no longer public. But you can check the wayback machine here: http://web.archive.org/web/20230202163724/https://pnpm.io/6.x/pnpx-cli
```suggestion
pnpm dlx storybook@latest upgrade --prerelease
```

File: code/renderers/vue3/template/stories/GlobalSetup.stories.js
Comment (JReinhold) [inline code/renderers/vue3/template/stories/GlobalSetup.stories.js:30]:
```suggestion
```

File: code/renderers/vue3/src/render.ts
Comment (JReinhold) [inline code/renderers/vue3/src/render.ts:35]: don't we need to remove setup functions again at some point, likely during teardown?
I'm thinking that if we don't, the setup functions will pollute each other, so if you open story A with its setup function, and then story B with its setup function, then when we render story B, we will run setup both for story A and B?

File: code/renderers/vue3/template/stories/GlobalSetup.stories.js
Comment (JReinhold) [inline code/renderers/vue3/template/stories/GlobalSetup.stories.js:7]: I see that the template stories here uses a pseudo-example of a setup function that emulates an i18n library, and a theming library. but I think they will be easier to understand and simpler if we just use a more explicit situation, that is more test-focused.
So I'd suggest getting rid of anything related to theme and i18n, and instead use simple setups that just adds text like "this text was added via a global setup function" and "this text was added via a setup function in render". then any maintainer can easily see that both texts are supposed to be shown, and that these template stories have nothing to do with i18n or theming at all.

**Review by JReinhold — CHANGES_REQUESTED** [review]
these changes look great, a few minor suggestions.

File: code/renderers/vue3/template/stories/GlobalSetup.stories.js
Comment (JReinhold) [inline code/renderers/vue3/template/stories/GlobalSetup.stories.js:33]:
```suggestion
    await expect(button).toHaveStyle('background-color: rgb(0, 128, 0)'); // <-- this provide primaryColor = green from .storybook/preview.ts
```

**Review by JReinhold — COMMENTED** [review]
Great! added a few minor comments, and there's still one question from previously you haven't resolved yet.

**Comment by chakAs3** [comment]
@JReinhold  is all green now.

**Review by JReinhold — COMMENTED** [review]
Changes look good from me, but I think @kasperpeulen should have a chance of reviewing it before it's merged.

You also haven't answered my comment here, you've just resolved it: https://github.com/storybookjs/storybook/pull/22170/files#r1173060551
can you tell me why my assumption is maybe wrong?

also i don't see why Kasper should review this as he does not have any context of this feature he was not in the Big Thread in Discord i guess @shilman is the one able to review this, we had a big discussion about this with other developers. it is tested and working fine for them it is been 3 weeks waiting for that. i don't know why we really move slowly for such trivial changes ?

File: code/renderers/vue3/src/render.ts
Comment (chakAs3) [reply code/renderers/vue3/src/render.ts:35]: @JReinhold this was not good implementation in previous version, and it has one comment by @shilman  that we should look for better implementation 
https://github.com/storybookjs/storybook/blob/303e39407653d0d0b7e23291dc01ed806f3e5f8a/code/renderers/vue3/template/stories/preview.js

**Comment by blowsie** [comment]
Yes we are still waiting for this to be merged. 
@JReinhold some more context can be found here.
https://discord.com/channels/486522875931656193/1090996147620950067

File: code/renderers/vue3/src/render.ts
Comment (chakAs3) [reply code/renderers/vue3/src/render.ts:35]:
```ts
// TODO: I'd like to be able to export rather than imperatively calling an imported function
// export const setup = (app) => {
//   app.component('GlobalButton', Button);
// };
```
so he really had the right concept which is making these functions available for other modules,  but the implementation is easy with vue, instead of exporting them you just have to add them to global Set or Array,  so at creation of Vue App you can iterate on this Set of functions ( i used Set to avoid duplication ) and run them

**Review by chakAs3 — COMMENTED** [review]
@blowsie do have any suggestion ? you can added your review if it seems fine for you then approve

**Comment by blowsie** [comment]
> @blowsie do have any suggestion ? you can added your review if it seems fine for you then approve

Sorry any suggestion for the code changes? Or is there an issue that needs addressing?

**Review by blowsie — APPROVED** [review]

**Comment by JReinhold** [comment]
@chakAs3 you'll notice that the comment you screenshotted comment has a "Pending" tag next to your name, which means that it was never published and is only visible by you. You probably wrote it as part of a review that you never published.

Either way, thanks for the elaboration. I didn't know that the `setup()` function was only meant to be invoked globally in `preview.js`, then it makes sense.

**Comment by chakAs3** [comment]
> @chakAs3 you'll notice that the comment you screenshotted comment has a "Pending" tag next to your name, which means that it was never published and is only visible by you. You probably wrote it as part of a review that you never published.
> 
> Either way, thanks for the elaboration. I didn't know that the `setup()` function was only meant to be invoked globally in `preview.js`, then it makes sense.

sorry i found out that was pending late.

**Review by kasperpeulen — APPROVED** [review]
Looks good to me, and thanks!

**Review by shilman — APPROVED** [review]
This looks good to me. I'd like @kasperpeulen to make sure this is well-tested, but we can do that after the fact since users are blocked on this. Thanks so much for putting it together. Great change!

**Comment by chakAs3** [comment]
> > this PR should be merged after merging #21956
> 
> @chakAs3 Is this still true? If so, we need to get that one reviewed & merged too?

it is already done the PR is merged

**Pull request merged**
